x-pack/plugins/fleet/server/services/artifacts/artifacts.ts

```diff
@@ -106,7 +106,7 @@ const generateArtifactBatches = (
 
   sortedArtifacts.forEach((artifact, index) => {
     const esArtifactResponse = esSearchHitToArtifact({
-      _id: uniqueIdFromArtifact(artifact),
+      _id: uniqueIdFromArtifact(artifacts[index]),
       _source: newArtifactToElasticsearchProperties(artifacts[index]),
```

Comment:
I think there is still a mismatch: we have `artifacts` and a sorted `sortedArtifacts`.

now lines 109 and 110 works with `artifacts[index]` which gets an artifact of the original, unsorted array, but below from line 112 we still use `artifact` from sorted array we iterate on.

I think that both line 109 and 110 should use `artifact`. what do you think?

Reply:
yeah make sense, what about `nonSortedEsArtifactResponse` ?

Reply:
@gergoabraham https://github.com/elastic/kibana/pull/160710/commits/af4805a0dc99ea4299a1dc600cfb23f90dfcd472

Reply:
after trying to understand it better, it looks like to me that building the batches based on the sorted order and generating the responses in the original order are two independent operations - maybe splitting them to a `forEach` and a `map` would ease the understanding of the code. what do you think?

Reply:
Yeah, that's right, but mi intention here was to avoid having two different loops and do it in a single one so we don't go through the array twice, thoughts?

Reply:
sorry, I didn't see your comment with the commit - thanks for changing the name!

I don't think that iterating twice makes such a difference, theoretically it's still O(n), only the handling of the second iteration gives a bit overhead, if the steps inside the iteration are the same.

however, in this specific case, the steps inside the iteration wouldn't be the same: now we need to push the items into `artifactsEsResponse`, which makes the iteration O(n^2) worst case, while using `map` would reduce this to O(n) - which is not perfectly true on the whole, because there is the push for the batches, too 😅 

anyway... without measurements I'm not sure which one is faster, I guess both are fast enough thanks to optimisations in node, so personally I'd go with the one that's easier to read and therefore easier to work with.

if you prefer having only one iterations, I think what really could help is to use a classic `for`, so the reader doesn't think from the first moment that the algorithm is working on `sortedArtifacts`, but sees immediately that there are more things happening in the iteration